instrumentation/vertx/vertx-sql-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/sql/QueryExecutorInstrumentation.java

```diff
@@ -0,0 +1,127 @@
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        isConstructor(), QueryExecutorInstrumentation.class.getName() + "$ConstructorAdvice");
+    transformer.applyAdviceToMethod(
+        namedOneOf("executeSimpleQuery", "executeExtendedQuery", "executeBatchQuery"),
+        QueryExecutorInstrumentation.class.getName() + "$QueryAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class ConstructorAdvice {
+
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void onExit(@Advice.This Object queryExecutor) {
+      // copy connection options from ThreadLocal to VirtualField
+      QueryExecutorUtil.setConnectOptions(queryExecutor, getSqlConnectOptions());
+    }
+  }
+
+  @SuppressWarnings("unused")
+  public static class QueryAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnter(
+        @Advice.This Object queryExecutor,
+        @Advice.AllArguments Object[] arguments,
+        @Advice.Local("otelCallDepth") CallDepth callDepth,
+        @Advice.Local("otelRequest") VertxSqlClientRequest otelRequest,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope) {
+      callDepth = CallDepth.forClass(queryExecutor.getClass());
+      if (callDepth.getAndIncrement() > 0) {
+        return;
+      }
+
+      // The parameter we need are in different positions, we are not going to have separate
+      // advices for all of them. The method gets the statement either as String or
+      // PreparedStatement, use the first argument that is either of these. PromiseInternal is
+      // always at the end of the argument list.
+      String sql = null;
+      PromiseInternal<?> promiseInternal = null;
+      for (Object argument : arguments) {
+        if (sql == null) {
+          if (argument instanceof String) {
+            sql = (String) argument;
+          } else if (argument instanceof PreparedStatement) {
+            sql = ((PreparedStatement) argument).sql();
+          }
+        } else if (argument instanceof PromiseInternal) {
+          promiseInternal = (PromiseInternal) argument;
+        }
+      }
+      if (sql == null || promiseInternal == null) {
+        return;
+      }
+
+      otelRequest =
+          new VertxSqlClientRequest(sql, QueryExecutorUtil.getConnectOptions(queryExecutor));
+      Context parentContext = currentContext();
+      if (!instrumenter().shouldStart(parentContext, otelRequest)) {
+        return;
+      }
+
+      context = instrumenter().start(parentContext, otelRequest);
+      scope = context.makeCurrent();
+      promiseInternal.context().localContextData().put(OTEL_REQUEST_KEY, otelRequest);
+      promiseInternal.context().localContextData().put(OTEL_CONTEXT_KEY, context);
+      promiseInternal.context().localContextData().put(OTEL_PARENT_CONTEXT_KEY, parentContext);
```

Comment:
Are these technically visible to the user? Assuming so, I'm not overly concerned about it, but just wanted to see if there's potential for user misuse.

Reply:
I believe user code has access to this context.